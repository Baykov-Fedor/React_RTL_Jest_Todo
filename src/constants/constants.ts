export const TODO_FIELDS = {
  ID: "id",
  VALUE: "value",
  COMPLETED: "completed",
} as const;

export const STORAGE_KEY = "my_todos";

export const TASK_TITLE = "task_title";

export const TASK_1 = {
  id: 1,
  value: "Brush teeth",
  completed: false,
};

export const TASK_2 = {
  id: 2,
  value: "Wash my car",
  completed: false,
};
