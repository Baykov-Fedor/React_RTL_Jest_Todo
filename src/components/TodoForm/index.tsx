import { useState } from "react";

import { TASK_TITLE, TODO_FIELDS } from "../../constants/constants";

import type { TodoFormProps } from "../typings";

const TodoForm = ({ addTodo }: TodoFormProps) => {
  const [taskTitle, setTaskTitle] = useState<string>("");

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setTaskTitle(event.target.value);
  };

  const handleKeyPress = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.key === "Enter") {
      const newTodo = {
        [TODO_FIELDS.ID]: Date.now(),
        [TODO_FIELDS.VALUE]: taskTitle,
        [TODO_FIELDS.COMPLETED]: false,
      };
      addTodo(newTodo);
      setTaskTitle("");
    }
  };

  return (
    <div className="input-field mt_2">
      <input
        placeholder="Type your task"
        className="main_input"
        name={TASK_TITLE}
        id={TASK_TITLE}
        type="text"
        onChange={handleChange}
        onKeyPress={handleKeyPress}
        value={taskTitle}
      />
      <label htmlFor={TASK_TITLE} className="active">
        Type here your "to do"
      </label>
    </div>
  );
};

export default TodoForm;
