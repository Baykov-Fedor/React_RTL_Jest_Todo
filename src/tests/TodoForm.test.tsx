import React from "react";
import { fireEvent, render } from "@testing-library/react";

import { TASK_1 } from "../constants/constants";
import TodoForm from "../components/TodoForm";
import userEvent from "@testing-library/user-event";

describe("Todo Form", () => {
  let handleTodoAdd: jest.Mock<any, any>;
  beforeEach(() => {
    handleTodoAdd = jest.fn();
  });

  it("Renders todo form", () => {
    const { asFragment } = render(<TodoForm addTodo={handleTodoAdd} />);
    expect(asFragment).toMatchSnapshot();
  });

  it("Working handler", () => {
    const { getByLabelText } = render(<TodoForm addTodo={handleTodoAdd} />);
    const input = getByLabelText('Type here your "to do"');
    expect(handleTodoAdd).toHaveBeenCalledTimes(0);
    fireEvent.keyPress(input, { key: "Enter", code: 13, charCode: 13 });
    expect(handleTodoAdd).toHaveBeenCalledTimes(1);
  });

  it("Setting and cleaning input", () => {
    const { getByLabelText } = render(<TodoForm addTodo={handleTodoAdd} />);
    const input = getByLabelText('Type here your "to do"');
    expect(input).toHaveValue("");
    userEvent.type(input, TASK_1.value);
    expect(input).toHaveValue(TASK_1.value);
    expect(handleTodoAdd).toHaveBeenCalledTimes(0);
    fireEvent.keyPress(input, { key: "Enter", code: 13, charCode: 13 });
    expect(handleTodoAdd).toHaveBeenCalledTimes(1);
    expect(input).toHaveValue("");
  });
});
