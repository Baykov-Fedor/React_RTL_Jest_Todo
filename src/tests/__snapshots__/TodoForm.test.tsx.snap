// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Form Renders todo form 1`] = `[Function]`;
